move-vm-support/src/ss58_address.rs

```diff
@@ -0,0 +1,145 @@
+//! SS58 address format converter for Substrate and Move accounts.
+
+use anyhow::{bail, Result};
+use blake2::{Blake2b512, Digest};
+use move_core_types::account_address::AccountAddress;
+
+// Substrate address prefix
+// Read more: https://docs.substrate.io/reference/address-formats/
+const SS58_PREFIX: &[u8] = b"SS58PRE";
+
+// Public key length in bytes
+const PUB_KEY_LEN: usize = 32;
+
+// Checksum length in bytes
+const CHECKSUM_LEN: usize = 2;
+
+// Blake2b512 hash length in bytes
+const HASH_LEN: usize = 64;
+
+// Maximum supported address type length in bytes
+const ADDR_TYPE_MAX_LEN: usize = 2;
+
+// Minimum supported address type length in bytes
+const ADDR_TYPE_MIN_LEN: usize = 1;
+
+// Maximum supported SS58 address length in bytes
+const SS58_MAX_LEN: usize = ADDR_TYPE_MAX_LEN + PUB_KEY_LEN + CHECKSUM_LEN;
+
+// Minimum supported SS58 address length in bytes
+const SS58_MIN_LEN: usize = ADDR_TYPE_MIN_LEN + PUB_KEY_LEN + CHECKSUM_LEN;
+
+/// Convert ss58 address string to Move address structure
+/// In case if such conversion is not possible, return error.
+/// ```
+/// use move_vm_support::ss58_address::ss58_to_move_address;
+/// let substrate_address = "gkNW9pAcCHxZrnoVkhLkEQtsLsW5NWTC75cdAdxAMs9LNYCYg";
+/// let move_address = ss58_to_move_address(substrate_address).unwrap();
+/// assert_eq!(
+///    "0x8EAF04151687736326C9FEA17E25FC5287613693C912909CB226AA4794F26A48",
+///   format!("{:#X}", move_address)
+/// );
+/// ```
+pub fn ss58_to_move_address(ss58: &str) -> Result<AccountAddress> {
+    // Decoded format: <addr_type>|<address>|<checksum>
+    //  Size in bytes:   1 or 2   |   32    |    2
+    let decoded_ss58 = bs58::decode(ss58).into_vec()?;
+
+    // Check if the length is valid and figure out the address type length.
+    let addr_type_len = match decoded_ss58.len() {
+        SS58_MIN_LEN => ADDR_TYPE_MIN_LEN,
+        SS58_MAX_LEN => ADDR_TYPE_MAX_LEN,
+        len => bail!("unsupported ss58 address length: {len}"),
+    };
+
+    let (type_and_addr, checksum) = &decoded_ss58.split_at(addr_type_len + PUB_KEY_LEN);
+
+    if *checksum != ss58_checksum(type_and_addr) {
+        bail!("invalid address checksum");
+    }
+
+    let (addr_type, address) = type_and_addr.split_at(addr_type_len);
+
+    // Sanity checks here
+    match addr_type_len {
+        ADDR_TYPE_MIN_LEN => {
+            if addr_type[0] > 63 && addr_type[0] < 128 {
+                bail!("invalid address length, address types from 64 to 127 are two bytes long");
+            }
+        }
+        ADDR_TYPE_MAX_LEN => {
+            if addr_type[0] < 64 {
+                bail!(
+                    "invalid address length, address types from 0 to 63 are exactly one byte long"
+                );
+            }
+        }
+        _ => unreachable!(),
```

Comment:
for the MIN condition, we can simpy check it if it is above 65 probably, since the range `(128..256)` is the undefined territory :D 

just my nit suggestion (we can use in the match arms):
```suggestion
        ADDR_TYPE_MIN_LEN if addr_type[0] > 64 => {
			bail!("invalid address type length, address types from 64 to 127 are two bytes long");
        }
        ADDR_TYPE_MAX_LEN if !(64..=128).contains(&addr_type[0]) => {
			bail!("invalid address length, address types from 0 to 63 are exactly one byte long");
        }
        _ => (),
```

Looking at the same code with the if-else isn't also looking bad:
```
        if addr_type_len == ADDR_TYPE_MIN_LEN && addr_type[0] > 64 => {
			bail!("...");
        }
        if addr_type_len == ADDR_TYPE_MAX_LEN && !(64..=128).contains(&addr_type[0]) => {
			bail!("...");
        }
```